src/analysis/souffle/tests/taint_with_logic/test_taint_with_auth_logic.dl

```diff
@@ -0,0 +1,119 @@
+)).
+
+// the label of "other_input" is the empty set (which makes it public).
+// so it has no label elements
+
+// Module 2: ImageSelector
+//     reads:
+//         image_detection_boxes
+//         image_selection   // image selection action from user
+//     writes:
+//         selected_image_id
+bindFlatParticle("ImageSelector", $FlatParticle(
+    /* reads */
+    $Cons("image_detection_boxes", $Cons("image_selection", $Nil())),
+    /* writes */
+    $Cons("selected_image_id", $Nil())
+)).
+
+//-----------------------------------------------------------------------------
+// Policy
+//-----------------------------------------------------------------------------
+
+// Assign owners to tags
+ownsTag("Alice", "raw_video_tag").
+ownsTag("Alice", "detected_images_tag").
+ownsTag("Alice", "image_selection_tag").
+ownsTag("Alice", "product_id_tag").
+ownsTag("ServiceProvider", "image_detection_model_tag").
+
+// Assign tags to initial inputs
+claimHasTag("sensor_data_packet.camera_feed", "raw_video_tag").
+// the access path, sensor_data_packet.video_resolution does not have sensitive 
+// data in this example -- it's just a configuration option that is 
+// uninteresting w.r.t privacy, so it has no tags (it's fully public / the empty set).
+claimHasTag("image_selection", "image_selection_tag").
+claimHasTag("detection_model", "image_detection_model_tag").
+
+// The label of the access path "image_detection_boxes" is downgraded from
+// {raw_video_tag, image_detection_model_tag} to {detected_images_tag}. This happens in two 
+// steps -- first the label is downgraded from {raw_video_tag, 
+// image_detection_tag} to {} (the empty set, which is totally public -- can go 
+// anywher). This downgrade happens with
+//      saysDowngrades(Alice, raw_video_tag, image_detection_boxes)
+// and
+//      saysDowngrades(ServiceProvider, image_detection_model_tag, 
+//            image_detection_boxes)
+// Then it is "upgraded" from {} to {product_id} because the label
+// {product_id} is more constrained. The upgrade happens with
+// 
+//      claimHasTag(image_detection_boxes, product_id).
+
+saysDowngrades("Alice", "image_detection_boxes", "raw_video_tag").
+saysDowngrades("ServiceProvider", "image_detection_boxes", 
+    "image_detection_model_tag").
+claimHasTag("image_detection_boxes", "image_selection_tag").
+
+// The output of the image selector is downgraded from 
+// {image_detection_model_tag} to {select
+saysDowngrades("Alice", "selected_image_id", "image_selection_tag").
+saysDowngrades("Alice", "selected_image_id", "detected_images_tag").
+claimHasTag("selected_image_id", "product_id_tag").
+
+
+//----------------------------------------------------------------------------- 
+// Policy Check
+//-----------------------------------------------------------------------------
+
+// Define a label for doing the check of the final output
+isIFCLabel("selected_image_id_expected_label").
+isMember("product_id_tag", "selected_image_id_expected_label").
```

Comment:
Is it correct that facts for `isMember` is manually specified. If so, is there value in defining `IFCLabel`? It seems like we might as well have multiple checks on the desired `tags` for `flowsTo`?

Reply:
I am confused about `separating accessPaths from labels`. Isn't it already the case?

Reply:
It is worth having some further discussion. To unblock you, it might help to be incremental here. Can we split the `IFCLabel`, `accessPathLabelFlowsTo`, etc. to a separate PR? In this PR, let us just introduce the `owns`, `downgrades`, `mayHaveTag`, `saysDowngrades`, etc. and merge it in.

Reply:
I think essentially because we have `claimHasTag(tag, path)` the path also acts like a label in a sense. If we only had `pathHasLabel(path, label)` and then used just used `isMember(tag, label)` the purpose of the two objects (paths and labels) would be more cleanly separated IMO. (It might look a bit [like this one](https://github.com/google-research/raksha/blob/61947009acc1e2b6312f80abd8d81d70cef7b730/datalog-prototyping/test_counterExampleLabelCheck.dl)).

I'm not really sure yet whether it would be an extra redundant layer or if it helps, but I want to do two things soon that might affect this:
 - I also would like to re-try separating this into 1) a label-check pass and 2) a label inference pass (using DFA)
 - Support confidentiality and integrity tags. Do downgrading with robust declassification / transparent endorsement + auth logic